WCSimFLOWER.cpp

```diff
@@ -345,56 +394,65 @@ void WCSimFLOWER::GetNEff()
+          nearbyHits++;
+      }//j
+
+      // PMTs at the top/bottom edge of the barrel only have five neighbours
+      if(abs(z) > fTopBottomDistanceLo && abs(z) < fTopBottomDistanceHi)
+        // TODO: deal with PMTs at the edge of the top/bottom
+        ratio = nearbyHits / 5.0;
+      else
+        ratio = nearbyHits / 8.0;
+
+      if (ratio == 0) {
+        occupancy = 1.0;
+      } else if (ratio < 1) {
+        occupancy = log(1 / (1-ratio)) / ratio; // from Poisson statistics
+      } else {
+        occupancy = 3.0;
+      }
     } else {
-      occupancy = 3.0;
+      // Ignore occupancy correction for 3" PMTs in mPMTs, since they are highly unlikely to receive multiple hits due to their small size.
+      occupancy = 1.0;
     }
 
     // correct for delayed hits (e.g. due to scattering)
-    lateHits = (fNMaxLong - fNMaxShort - (fNWorkingPMTs * fDarkRate * fShortDuration)) / fNMaxShort;
+    lateHits = (fNMaxLong - fNMaxShort - (fNWorkingPMTs * fDarkRate * fShortDuration)
+                                       - (fNWorkingPMTs2 * fDarkRate2 * fShortDuration)) / fNMaxShort;
 
     // substract dark noise hits
-    darkNoise = (fNWorkingPMTs * fDarkRate * fShortDuration) / fNMaxShort;
+    darkNoise = (fNWorkingPMTs * fDarkRate * fShortDuration + fNWorkingPMTs2 * fDarkRate2 * fShortDuration) / fNMaxShort;
 
     // calculate effective coverage to correct for photoCoverage
     // this depends on angle of incidence, see Fig. 4.5 (left) of http://www-sk.icrr.u-tokyo.ac.jp/sk/_pdf/articles/2016/doc_thesis_naknao.pdf
     // TODO: take into account azimuthal dependence as show in Fig. 4.5 (right); currently assumes phi=0
-    float dotProduct = pmt.GetOrientation(0)*(fVertex[0] - x) +
-      pmt.GetOrientation(1)*(fVertex[1] - y) +
-      pmt.GetOrientation(2)*(fVertex[2] - z);
+    float dotProduct = pmt.GetOrientation(0) * (fVertex[0] - x) +
+                       pmt.GetOrientation(1) * (fVertex[1] - y) +
+                       pmt.GetOrientation(2) * (fVertex[2] - z);
     float theta = acos( dotProduct / fDistanceShort[i]) * 180 / 3.14159265;
     if (!((theta >= 0.) and (theta <= 89.99))) theta = 0; // we have apparently mis-reconstructed the vertex, so let's set ...
                                                           // ... the coverage to the most likely value of 0.4 (i.e. theta < 40 degrees)
     
-    photoCoverage = 1 / WCSimFLOWER::fEffCoverages[int(theta/10)];
+    photoCoverage = 1 / WCSimFLOWER::fEffCoverages[int(theta/10)]; // 1 over photocoverage, assuming SuperK or HyperK_40perCent
     if (fDetector == kHyperK20)
-      photoCoverage *= 38448/float(19462); // ratio of number of PMTs is not exactly 2
+      photoCoverage *= 38448/float(19462); // ratio of number of PMTs
+    if (fDetector == kHyperK20BnL10mPMT)
+      photoCoverage *= 38448/float(19462); // ratio of number of B&L PMTs
+      photoCoverage /= 1.22; // 10k mPMT modules (19 PMTs, 3" diameter) have 22% the area of 19462 20-inch PMTs
```

Comment:
Isn't this a syntax issue? mPMT are only in kHyperK20BnL10mPMT, so it should rather be:
```

    if (fDetector == kHyperK20BnL10mPMT) {
      photoCoverage *= 38448/float(19462); // ratio of number of B&L PMTs
      photoCoverage /= 1.22; // 10k mPMT modules (19 PMTs, 3" diameter) have 22% the area of 19462 20-inch PMTs
    }
```

Reply:
Yep, that’s a syntax issue. (Can you tell I’m more used to Python? 😅)
I’ll also change the PMT number; though we’ll have to revisit that for all geometries once hybrid & main WCSim are merged and their geometries agree.
Many thanks for both comments!

Reply:
I was guessing so (for python usage) :D 
Thanks for implementing them!